File: templates/problems/08_tries/02_word_dictionary/solution.go
package word_dictionary

// SolveWordDictionary supports adding words and searching with '.' wildcards.
type SolveWordDictionary struct {
	Root *TrieNode
}

// NewSolveWordDictionary creates and returns a new SolveWordDictionary.
func NewSolveWordDictionary() SolveWordDictionary {
	return SolveWordDictionary{Root: &TrieNode{Children: make(map[byte]*TrieNode)}}
}

// AddWord adds a word to the dictionary.
// Time: O(m), Space: O(m).
func (wd *SolveWordDictionary) AddWord(word string) {
	node := wd.Root
	for i := 0; i < len(word); i++ {
		c := word[i]
		if node.Children[c] == nil {
			node.Children[c] = &TrieNode{Children: make(map[byte]*TrieNode)}
		}
		node = node.Children[c]
	}
	node.IsEnd = true
}

// Search returns true if the word matches any string in the dictionary.
// '.' can match any single character.
// Time: O(26^d * m) worst case where d is number of dots, Space: O(m).
func (wd *SolveWordDictionary) Search(word string) bool {
	return searchNode(wd.Root, word, 0)
}

func searchNode(node *TrieNode, word string, i int) bool {
	if i == len(word) {
		return node.IsEnd
	}
	c := word[i]
	if c == '.' {
		for _, child := range node.Children {
			if searchNode(child, word, i+1) {
				return true
			}
		}
		return false
	}
	child := node.Children[c]
	if child == nil {
		return false
	}
	return searchNode(child, word, i+1)
}

File: templates/problems/08_tries/02_word_dictionary/python/solution.py
from __future__ import annotations


class _SolveWDNode:
    __slots__ = ("children", "is_end")

    def __init__(self):
        self.children: dict[str, _SolveWDNode] = {}
        self.is_end = False


class SolveWordDictionary:
    """Supports adding words and searching with '.' wildcards."""

    def __init__(self):
        self.root = _SolveWDNode()

    def add_word(self, word: str) -> None:
        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = _SolveWDNode()
            node = node.children[c]
        node.is_end = True

    def search(self, word: str) -> bool:
        return _search_node(self.root, word, 0)


def _search_node(node: _SolveWDNode, word: str, i: int) -> bool:
    if i == len(word):
        return node.is_end
    c = word[i]
    if c == ".":
        for child in node.children.values():
            if _search_node(child, word, i + 1):
                return True
        return False
    child = node.children.get(c)
    if child is None:
        return False
    return _search_node(child, word, i + 1)
